[Inference API] Deprecate elser service

maxhniebergall opened the pull request:
This change adds a deprecation warning for the elser service on `Put` and `Infer` and copies the elser service files so we can simply delete the elser service after 8.16

This change maintains BWC with the elser service.

Commit:
merging

Commit:
copy elser service files into elasticsearch service

Commit:
copy elser service files into elasticsearch service

Commit:
Add deprecation log message for elser service

Commit:
Add deprecation log message for elser service

Commit:
improve deprecation warning

Commit:
change elasticsearch internal service elser case to use elser model

Commit:
switch elasticsearch elser tests to use elasticsearch elser

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

elasticsearchmachine commented:
Hi @maxhniebergall, I've created a changelog YAML for you. *Note* that since this PR is labelled `>deprecation`, you need to update the changelog YAML to fill out the extended information sections.

maxhniebergall committed:
Update docs/changelog/113216.yaml

davidkyle reviewed (commented):
Now that the elasticsearch service can run the elser model can we remove the elser service in this PR? The elasticsearch service can be an alias to the elser service, adding some logic to the `InferenceServiceRegistry` lookup code would simply return the elasticsearch service when elser is requested.

Commit:
alias elser service to elasticsearch

Commit:
delete elser service package now that elasticsearch service supports it and has aliased it

Commit:
Add deprecation warning to infer API for elser

Commit:
Add deprecation warning to infer API for elser

Commit:
Fix accidentally introduced NPE and retain BWC support for null model ID (with deprecation message)

maxhniebergall committed:
change "area" to "REST API" because "Machine Learning" isn't an option for deprecation

Commit:
change elser literals to static variable

Commit:
change Put and Elasticsearch Internal service

 to pass the service name if it is elser or elasticsearch this will allow the elasticsearch service to maintain BWC for null model IDs if the service was elser.

Commit:
fix up tests to match new elasticsearch service semantics regarding elser.

Commit:
Move passing of service name

Commit:
add persistence for elser models in elasticsearch

maxhniebergall commented:
@elasticmachine merge upstream

elasticmachine commented:
merge conflict between base and head

Commit:
fix merge conflicts

davidkyle reviewed (approved):
LGTM

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceAction.java:
```diff
@@ -75,6 +80,15 @@ protected void doExecute(Task task, InferenceAction.Request request, ActionListe
                 );
                 return;
             }
+            if (service.get().name().equals(OLD_ELSER_SERVICE_NAME)) {
+                DEPRECATION_LOGGER.warn(
+                    DeprecationCategory.API,
+                    "inference_api_elser_service",
+                    "The [{}] service is deprecated and will be removed in a future release. Use the [{}] service instead.",
+                    OLD_ELSER_SERVICE_NAME,
+                    ElasticsearchInternalService.NAME
+                );
+            }
```
```suggestion
```

Commit:
Merge branch 'main' of github.com:elastic/elasticsearch into mergeElserService

# Conflicts:
#	x-pack/plugin/inference/src/internalClusterTest/java/org/elasticsearch/xpack/inference/integration/ModelRegistryIT.java
#	x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportPutInferenceModelAction.java
#	x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/BaseElasticsearchInternalService.java
#	x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/ElasticsearchInternalService.java
#	x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elser/ElserInternalService.java
#	x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/elasticsearch/ElasticsearchInternalServiceTests.java
#	x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/elser/ElserInternalServiceTests.java

Commit:
fix merge

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 113216`
